operators/pkg/controller/kibana/securesettings/user_secret.go

```diff
@@ -0,0 +1,123 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package securesettings
+
+import (
+	"fmt"
+
+	commonv1alpha1 "github.com/elastic/cloud-on-k8s/operators/pkg/apis/common/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/apis/kibana/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/events"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/finalizer"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/volume"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/common/watches"
+	kbvolume "github.com/elastic/cloud-on-k8s/operators/pkg/controller/kibana/volume"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+)
+
+// secureSettingsVolume creates a volume from the optional user-provided secure settings secret.
+//
+// Secure settings are provided by the user in the Kibana Spec through a secret reference.
+// This secret is mounted into Kibana pods for secure settings to be injected into Kibana keystore.
+// The user-provided secret is watched to reconcile on any change.
+// The user secret resource version is returned along with the volume, so that
+// any change in the user secret leads to pod rotation.
+func secureSettingsVolume(
+	c k8s.Client,
+	recorder record.EventRecorder,
+	watches watches.DynamicWatches,
+	kb v1alpha1.Kibana,
+) (*volume.SecretVolume, string, error) {
+	// setup (or remove) watches for the user-provided secret to reconcile on any change
+	userSecretRef := kb.Spec.SecureSettings
+	err := watchSecureSettings(watches, userSecretRef, k8s.ExtractNamespacedName(&kb))
+	if err != nil {
+		return nil, "", err
+	}
+
+	if userSecretRef == nil {
+		// no secure settings secret specified
+		return nil, "", nil
+	}
+
+	// retrieve the secret referenced by the user in the Kibana namespace
+	userSecret, exists, err := retrieveUserSecret(c, kb, recorder, kb.Namespace, userSecretRef.SecretName)
+	if err != nil {
+		return nil, "", err
+	}
+	if !exists {
+		// secret does not exist (yet): no volume to mount
```

Comment:
Here, the user defined a `secureSettings.secretName` and the secret does not exist (yet). 
IIUC we act as if there is not secret and continue and in another next reconciliation loop we will find the secret and then recreate the pod?

Reply:
Exactly. We do log and emit an event to signal this to the user though.

Reply:
Interesting behaviour. I would think it is more of a user mistake and we should not start Kibana if a secret is referenced but does not exist. But I'm fine with your proposal.

Reply:
Hmm maybe you're right?

The same behaviour is implemented for the Elasticsearch keystore though: we don't stop the reconciliation if the keystore secret ref is unspecified. What if you add a wrong secret ref once the cluster is already running? If we'd return early from the reconciliation loop for that reason, we'd stop managing other important parts of the cluster in the rest of the reconciliation. Maybe we could separate initial resource creation from update but that becomes a bit complicated.

Reply:
Makes sense. Let's keep it that way for now.